### PR TITLE
{2023.06}[foss/2022a] ESPResSo v4.2.1

### PR DESCRIPTION
**Pull request by maxim-masterov**
Add ESPResSo

Depends on:
* #310

edit (by @boegel, 2023-10-10):

```
33 out of 78 required modules missing:

* Szip/2.1.1-GCCcore-11.3.0 (Szip-2.1.1-GCCcore-11.3.0.eb)
* gzip/1.12-GCCcore-11.3.0 (gzip-1.12-GCCcore-11.3.0.eb)
* lz4/1.9.3-GCCcore-11.3.0 (lz4-1.9.3-GCCcore-11.3.0.eb)
* zstd/1.5.2-GCCcore-11.3.0 (zstd-1.5.2-GCCcore-11.3.0.eb)
* GSL/2.7-GCC-11.3.0 (GSL-2.7-GCC-11.3.0.eb)
* ICU/71.1-GCCcore-11.3.0 (ICU-71.1-GCCcore-11.3.0.eb)
* Bison/3.8.2-GCCcore-11.3.0 (Bison-3.8.2-GCCcore-11.3.0.eb)
* Ninja/1.10.2-GCCcore-11.3.0 (Ninja-1.10.2-GCCcore-11.3.0.eb)
* Pint/0.22-GCCcore-11.3.0 (Pint-0.22-GCCcore-11.3.0.eb)
* Mako/1.2.0-GCCcore-11.3.0 (Mako-1.2.0-GCCcore-11.3.0.eb)
* OpenPGM/5.2.122-GCCcore-11.3.0 (OpenPGM-5.2.122-GCCcore-11.3.0.eb)
* Meson/0.62.1-GCCcore-11.3.0 (Meson-0.62.1-GCCcore-11.3.0.eb)
* BeautifulSoup/4.10.0-GCCcore-11.3.0 (BeautifulSoup-4.10.0-GCCcore-11.3.0.eb)
* libsodium/1.0.18-GCCcore-11.3.0 (libsodium-1.0.18-GCCcore-11.3.0.eb)
* ZeroMQ/4.3.4-GCCcore-11.3.0 (ZeroMQ-4.3.4-GCCcore-11.3.0.eb)
* libxslt/1.1.34-GCCcore-11.3.0 (libxslt-1.1.34-GCCcore-11.3.0.eb)
* lxml/4.9.1-GCCcore-11.3.0 (lxml-4.9.1-GCCcore-11.3.0.eb)
* IPython/8.5.0-GCCcore-11.3.0 (IPython-8.5.0-GCCcore-11.3.0.eb)
* libunwind/1.6.2-GCCcore-11.3.0 (libunwind-1.6.2-GCCcore-11.3.0.eb)
* Boost.MPI/1.79.0-gompi-2022a (Boost.MPI-1.79.0-gompi-2022a.eb)
* HDF5/1.12.2-gompi-2022a (HDF5-1.12.2-gompi-2022a.eb)
* LLVM/14.0.3-GCCcore-11.3.0 (LLVM-14.0.3-GCCcore-11.3.0.eb)
* libiconv/1.17-GCCcore-11.3.0 (libiconv-1.17-GCCcore-11.3.0.eb)
* Doxygen/1.9.4-GCCcore-11.3.0 (Doxygen-1.9.4-GCCcore-11.3.0.eb)
* libpng/1.6.37-GCCcore-11.3.0 (libpng-1.6.37-GCCcore-11.3.0.eb)
* Brotli/1.0.9-GCCcore-11.3.0 (Brotli-1.0.9-GCCcore-11.3.0.eb)
* freetype/2.12.1-GCCcore-11.3.0 (freetype-2.12.1-GCCcore-11.3.0.eb)
* fontconfig/2.14.0-GCCcore-11.3.0 (fontconfig-2.14.0-GCCcore-11.3.0.eb)
* X11/20220504-GCCcore-11.3.0 (X11-20220504-GCCcore-11.3.0.eb)
* libdrm/2.4.110-GCCcore-11.3.0 (libdrm-2.4.110-GCCcore-11.3.0.eb)
* libglvnd/1.4.0-GCCcore-11.3.0 (libglvnd-1.4.0-GCCcore-11.3.0.eb)
* Mesa/22.0.3-GCCcore-11.3.0 (Mesa-22.0.3-GCCcore-11.3.0.eb)
* ESPResSo/4.2.1-foss-2022a (ESPResSo-4.2.1-foss-2022a.eb)
```

### CODE REVIEW

**Comment by maxim-masterov** [comment]
```
58 out of 78 required modules missing:

* UnZip/6.0-GCCcore-11.3.0 (UnZip-6.0-GCCcore-11.3.0.eb)
* cURL/7.83.0-GCCcore-11.3.0 (cURL-7.83.0-GCCcore-11.3.0.eb)
* libarchive/3.6.1-GCCcore-11.3.0 (libarchive-3.6.1-GCCcore-11.3.0.eb)
* CMake/3.23.1-GCCcore-11.3.0 (CMake-3.23.1-GCCcore-11.3.0.eb)
* Szip/2.1.1-GCCcore-11.3.0 (Szip-2.1.1-GCCcore-11.3.0.eb)
* Eigen/3.4.0-GCCcore-11.3.0 (Eigen-3.4.0-GCCcore-11.3.0.eb)
* gzip/1.12-GCCcore-11.3.0 (gzip-1.12-GCCcore-11.3.0.eb)
* GMP/6.2.1-GCCcore-11.3.0 (GMP-6.2.1-GCCcore-11.3.0.eb)
* lz4/1.9.3-GCCcore-11.3.0 (lz4-1.9.3-GCCcore-11.3.0.eb)
* libffi/3.4.2-GCCcore-11.3.0 (libffi-3.4.2-GCCcore-11.3.0.eb)
* Tcl/8.6.12-GCCcore-11.3.0 (Tcl-8.6.12-GCCcore-11.3.0.eb)
* zstd/1.5.2-GCCcore-11.3.0 (zstd-1.5.2-GCCcore-11.3.0.eb)
* SQLite/3.38.3-GCCcore-11.3.0 (SQLite-3.38.3-GCCcore-11.3.0.eb)
* GSL/2.7-GCC-11.3.0 (GSL-2.7-GCC-11.3.0.eb)
* Python/3.10.4-GCCcore-11.3.0-bare (Python-3.10.4-GCCcore-11.3.0-bare.eb)
* ICU/71.1-GCCcore-11.3.0 (ICU-71.1-GCCcore-11.3.0.eb)
* Rust/1.60.0-GCCcore-11.3.0 (Rust-1.60.0-GCCcore-11.3.0.eb)
* Bison/3.8.2-GCCcore-11.3.0 (Bison-3.8.2-GCCcore-11.3.0.eb)
* FFTW/3.3.10-GCC-11.3.0 (FFTW-3.3.10-GCC-11.3.0.eb)
* BLIS/0.9.0-GCC-11.3.0 (BLIS-0.9.0-GCC-11.3.0.eb)
* git/2.36.0-GCCcore-11.3.0-nodocs (git-2.36.0-GCCcore-11.3.0-nodocs.eb)
* Python/3.10.4-GCCcore-11.3.0 (Python-3.10.4-GCCcore-11.3.0.eb)
* hypothesis/6.46.7-GCCcore-11.3.0 (hypothesis-6.46.7-GCCcore-11.3.0.eb)
* pybind11/2.9.2-GCCcore-11.3.0 (pybind11-2.9.2-GCCcore-11.3.0.eb)
* Ninja/1.10.2-GCCcore-11.3.0 (Ninja-1.10.2-GCCcore-11.3.0.eb)
* Pint/0.22-GCCcore-11.3.0 (Pint-0.22-GCCcore-11.3.0.eb)
* Mako/1.2.0-GCCcore-11.3.0 (Mako-1.2.0-GCCcore-11.3.0.eb)
* OpenPGM/5.2.122-GCCcore-11.3.0 (OpenPGM-5.2.122-GCCcore-11.3.0.eb)
* Meson/0.62.1-GCCcore-11.3.0 (Meson-0.62.1-GCCcore-11.3.0.eb)
* BeautifulSoup/4.10.0-GCCcore-11.3.0 (BeautifulSoup-4.10.0-GCCcore-11.3.0.eb)
* libsodium/1.0.18-GCCcore-11.3.0 (libsodium-1.0.18-GCCcore-11.3.0.eb)
* ZeroMQ/4.3.4-GCCcore-11.3.0 (ZeroMQ-4.3.4-GCCcore-11.3.0.eb)
* libxslt/1.1.34-GCCcore-11.3.0 (libxslt-1.1.34-GCCcore-11.3.0.eb)
* make/4.3-GCCcore-11.3.0 (make-4.3-GCCcore-11.3.0.eb)
* lxml/4.9.1-GCCcore-11.3.0 (lxml-4.9.1-GCCcore-11.3.0.eb)
* OpenBLAS/0.3.20-GCC-11.3.0 (OpenBLAS-0.3.20-GCC-11.3.0.eb)
* IPython/8.5.0-GCCcore-11.3.0 (IPython-8.5.0-GCCcore-11.3.0.eb)
* FlexiBLAS/3.2.0-GCC-11.3.0 (FlexiBLAS-3.2.0-GCC-11.3.0.eb)
* libunwind/1.6.2-GCCcore-11.3.0 (libunwind-1.6.2-GCCcore-11.3.0.eb)
* gompi/2022a (gompi-2022a.eb)
* FFTW.MPI/3.3.10-gompi-2022a (FFTW.MPI-3.3.10-gompi-2022a.eb)
* ScaLAPACK/2.2.0-gompi-2022a-fb (ScaLAPACK-2.2.0-gompi-2022a-fb.eb)
* Boost.MPI/1.79.0-gompi-2022a (Boost.MPI-1.79.0-gompi-2022a.eb)
* HDF5/1.12.2-gompi-2022a (HDF5-1.12.2-gompi-2022a.eb)
* foss/2022a (foss-2022a.eb)
* SciPy-bundle/2022.05-foss-2022a (SciPy-bundle-2022.05-foss-2022a.eb)
* LLVM/14.0.3-GCCcore-11.3.0 (LLVM-14.0.3-GCCcore-11.3.0.eb)
* libiconv/1.17-GCCcore-11.3.0 (libiconv-1.17-GCCcore-11.3.0.eb)
* Doxygen/1.9.4-GCCcore-11.3.0 (Doxygen-1.9.4-GCCcore-11.3.0.eb)
* libpng/1.6.37-GCCcore-11.3.0 (libpng-1.6.37-GCCcore-11.3.0.eb)
* Brotli/1.0.9-GCCcore-11.3.0 (Brotli-1.0.9-GCCcore-11.3.0.eb)
* freetype/2.12.1-GCCcore-11.3.0 (freetype-2.12.1-GCCcore-11.3.0.eb)
* fontconfig/2.14.0-GCCcore-11.3.0 (fontconfig-2.14.0-GCCcore-11.3.0.eb)
* X11/20220504-GCCcore-11.3.0 (X11-20220504-GCCcore-11.3.0.eb)
* libdrm/2.4.110-GCCcore-11.3.0 (libdrm-2.4.110-GCCcore-11.3.0.eb)
* libglvnd/1.4.0-GCCcore-11.3.0 (libglvnd-1.4.0-GCCcore-11.3.0.eb)
* Mesa/22.0.3-GCCcore-11.3.0 (Mesa-22.0.3-GCCcore-11.3.0.eb)
* ESPResSo/4.2.1-foss-2022a (ESPResSo-4.2.1-foss-2022a.eb)
```

**Comment by boegel** [comment]
@maxim-masterov #310 should get done first?

**Comment by boegel** [comment]
For both jobs: `Failed to get data for PR #18870`, so trying again...

bot: build repo:eessi-2023.06-software arch:x86_64/generic
bot: build repo:eessi-2023.06-software arch:aarch64/neoverse_v1

**Comment by boegel** [comment]
Wrong Python is being used:

```
ModuleNotFoundError: No module named 'numpy.core._multiarray_umath'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/cvmfs/pilot.eessi-hpc.org/versions/2023.06/software/linux/x86_64/generic/software/SciPy-bundle/2022.05-foss-2022a/lib/python3.10/site-packages/numpy/__init__.py", line 144, in <module>
    from . import core
  File "/cvmfs/pilot.eessi-hpc.org/versions/2023.06/software/linux/x86_64/generic/software/SciPy-bundle/2022.05-foss-2022a/lib/python3.10/site-packages/numpy/core/__init__.py", line 49, in <module>
    raise ImportError(msg)
ImportError:

IMPORTANT: PLEASE READ THIS FOR ADVICE ON HOW TO SOLVE THIS ISSUE!

Importing the numpy C-extensions failed. This error can happen for
many reasons, often due to issues with your setup or how NumPy was
installed.

We have compiled some common reasons and troubleshooting tips at:

    https://numpy.org/devdocs/user/troubleshooting-importerror.html

Please note and check the following:

  * The Python version is: Python3.11 from "/cvmfs/pilot.eessi-hpc.org/versions/2023.06/compat/linux/x86_64/usr/lib/python-exec/python3.11/python3"
  * The NumPy version is: "1.22.3"
```

May need to add additional CMake options like we did for VTK in https://github.com/easybuilders/easybuild-easyconfigs/pull/16741 ?

**Comment by maxim-masterov** [comment]
Incorporated changes in PR [18963](https://github.com/easybuilders/easybuild-easyconfigs/pull/18963)

File: eessi-2023.06-eb-4.8.1-2022a.yml [hunk 1]
```diff
@@ -5,5 +5,8 @@ easyconfigs:
       options:
         from-pr: 18870
   - foss-2022a
+  - ESPResSo-4.2.1-foss-2022a:
+      options:
+        from-pr: 18963
   - SciPy-bundle-2022.05-foss-2022a
-  - BAMM-2.5.0-foss-2022a.eb
\ No newline at end of file
+  - BAMM-2.5.0-foss-2022a.eb
```
Comment (boegel) [inline eessi-2023.06-eb-4.8.1-2022a.yml:12]: @maxim-masterov Let's try and preserve order in which installations were done here, so:

```suggestion
  - SciPy-bundle-2022.05-foss-2022a
  - BAMM-2.5.0-foss-2022a.eb
  - ESPResSo-4.2.1-foss-2022a:
      options:
        from-pr: 18963

**Review by boegel — CHANGES_REQUESTED** [review]

**Comment by boegel** [comment]
Let's take this for a spin, to verify whether the changes in https://github.com/easybuilders/easybuild-easyconfigs/pull/18963 are sufficient to fix the problem with picking up the Python dependency...

bot: build repo:eessi-2023.06-software arch:x86_64/generic
bot: build repo:eessi-2023.06-software arch:x86_64/intel/haswell
bot: build repo:eessi-2023.06-software arch:aarch64/neoverse_v1

**Comment by boegel** [comment]
Build still fails, same problem, so the `-DPython3_EXECUTABLE` being added is not sufficient to avoid that CMake picks up the Python 3.11 from the compat layer...

**Comment by boegel** [comment]
https://github.com/easybuilders/easybuild-easyconfigs/pull/18963 has been updated to use `-DPYTHON_EXECUTABLE` instead of `-DPython3_EXECUTABLE` (pro tip: don't use `-DPython_EXECUTABLE`...), so let's try the build again.

When doing a manual test the build worked fine, but I ran into a single failing test due to timeout:

```
99% tests passed, 1 tests failed out of 201

Label Time Summary:
gpu             = 240.37 sec*proc (56 tests)
long            = 152.77 sec*proc (13 tests)
parallel        = 763.71 sec*proc (45 tests)
parallel_odd    =   4.05 sec*proc (1 test)

Total Test time (real) = 702.04 sec

The following tests FAILED:
        164 - oif_volume_conservation (Timeout)
```

```
164/201 Test #164: oif_volume_conservation .......................................***Timeout 300.03 sec
```

@jngrad mentioned that this could be a flaky test, and that there are known issues with MPI callbacks in ESPResSo v4.2.1 to may cause hangs, which could be the cause here...

**Comment by boegel** [comment]
bot: build repo:eessi-2023.06-software arch:x86_64/generic
bot: build repo:eessi-2023.06-software arch:x86_64/intel/haswell
bot: build repo:eessi-2023.06-software arch:x86_64/intel/skylake_avx512
bot: build repo:eessi-2023.06-software arch:x86_64/amd/zen2
bot: build repo:eessi-2023.06-software arch:x86_64/amd/zen3
bot: build repo:eessi-2023.06-software arch:aarch64/generic
bot: build repo:eessi-2023.06-software arch:aarch64/neoverse_n1
bot: build repo:eessi-2023.06-software arch:aarch64/neoverse_v1

**Comment by boegel** [comment]
Particular tests are known to time out frequently, like `virtual_sites_relative`, see https://github.com/espressomd/espresso/issues/4639, so we'll probably need to ignore failures in the test step for ESPResSo v4.2.1 for the time being

**Comment by boegel** [comment]
bot: build repo:eessi-2023.06-software arch:aarch64/neoverse_v1

**Comment by boegel** [comment]
Failing tests in ESPResSo test suite on `aarch64/neoverse_v1`

In job `8207`:

```
97% tests passed, 7 tests failed out of 201
...
The following tests FAILED:
          4 - test_checkpoint__therm_lb__p3m_cpu__lj__lb_cpu_ascii (Timeout)
         76 - constant_pH_stats (Timeout)
         99 - hybrid_decomposition (Timeout)
        101 - integrator_npt_stats (Timeout)
        124 - collision_detection (Timeout)
        151 - thermostats_anisotropic (Timeout)
```

In job `8209`:

```
99% tests passed, 2 tests failed out of 201
...
The following tests FAILED:
         76 - constant_pH_stats (Failed)
        101 - integrator_npt_stats (Timeout)
```

For some reason it seems more likely to manifest on `aarch64/neoverse_v1`, since there were no test failures/timeouts in the builds for all other CPU targets, but that could be dumb luck (or there could be a link with #325).

I've opened #363 as a tracker issue, and will add a hook here so we can ignore failing tests if they happen for now, since we got feedback from @jngrad who's actively involved with the development of ESPResSo that some tests are known to hang occasionally due to problems in the ESPResSo codebase.

**Comment by jngrad** [comment]
Here are all statistical tests: `mass-and-rinertia_per_particle`, `rotational-diffusion-aniso`, `integrator_npt_stats.py`, `constant_pH_stats`, `langevin_thermostat_stats`, `brownian_dynamics_stats.py`, `dpd_stats`, `stokesian_dynamics`.

They are known to take a large amount of time on our CI pipelines, because we run them concurrently and max out the host machine CPU resource usage via MPI oversubscription, so that hyperthreaded cores are fully used. This makes their runtime fluctuate wildly, with a negative feedback loop since they compete against one another for the same resources (e.g. if one test times out, there is a very good chance another unrelated test will time out too). More details can be found in espressomd/espresso#3883.

Having said that, you don't seem to run these tests concurrently, so your CI pipelines should **not** be experiencing the issue I just described. Maybe there is a deeper issue in ESPResSo's MPI code, unfortunately timeout information alone is not sufficient for me to investigate an MPI issue.

**Comment by boegel** [comment]
> Here are all statistical tests: `mass-and-rinertia_per_particle`, `rotational-diffusion-aniso`, `integrator_npt_stats.py`, `constant_pH_stats`, `langevin_thermostat_stats`, `brownian_dynamics_stats.py`, `dpd_stats`, `stokesian_dynamics`.
> 
> They are known to take a large amount of time on our CI pipelines, because we run them concurrently and max out the host machine CPU resource usage via MPI oversubscription, so that hyperthreaded cores are fully used. This makes their runtime fluctuate wildly, with a negative feedback loop since they compete against one another for the same resources (e.g. if one test times out, there is a very good chance another unrelated test will time out too). More details can be found in [espressomd/espresso#3883](https://github.com/espressomd/espresso/issues/3883).
> 
> Having said that, you don't seem to run these tests concurrently, so your CI pipelines should **not** be experiencing the issue I just described. Maybe there is a deeper issue in ESPResSo's MPI code, unfortunately timeout information alone is not sufficient for me to investigate an MPI issue.

@jngrad Do you mind re-posting that in #363 ?

**Comment by ocaisa** [comment]
@boegel Do we have a false postive here? The only errors are related to finding
```
found message matching ERROR: 
```
and I suspect this is related to the tests that we ignore.

**Comment by boegel** [comment]
The `bot/build.sh` script is confused by the known issues YAML file:

```
>> Still no module for EasyBuild veessi-2023.06-known-issues.yml, trying with easyconfig EasyBuild-eessi-2023.06-known-issues.yml.eb...
ERROR: No easyconfig found for EasyBuild veessi-2023.06-known-issues.yml
```

I'll see why that happens, and how to fix it...

**Comment by boegel** [comment]
Problem caused by the "known issues" YAML file is fixed in 8779e8b.

I'll re-trigger the builds (will take 1-1.5h again, but fine, I don't want to cut corners here).

bot: build repo:eessi-2023.06-software arch:x86_64/generic
bot: build repo:eessi-2023.06-software arch:x86_64/intel/haswell
bot: build repo:eessi-2023.06-software arch:x86_64/intel/skylake_avx512
bot: build repo:eessi-2023.06-software arch:x86_64/amd/zen2
bot: build repo:eessi-2023.06-software arch:x86_64/amd/zen3
bot: build repo:eessi-2023.06-software arch:aarch64/generic
bot: build repo:eessi-2023.06-software arch:aarch64/neoverse_n1
bot: build repo:eessi-2023.06-software arch:aarch64/neoverse_v1

**Review by boegel — APPROVED** [review]

**Pull request merged**